In [1]:
# Step 1. Import the required libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import balanced_accuracy_score
from imblearn.pipeline import Pipeline as ImbPipeline
from imblearn.over_sampling import SMOTE

In [2]:
# Step 2. Download and initial review of data
train_data = pd.read_csv("final_proj_data.csv", index_col=0)
train_data.head()

test_data = pd.read_csv("final_proj_test.csv", index_col=0)
test_data.head()

Var2  Var3  Var4  Var5  ...  Var227         Var228  Var229  Var230
Var1                          ...                                       
NaN    NaN   NaN   NaN   NaN  ...    ZI9m  ib5G6X1eUxUn6    am7c     NaN
NaN    NaN   NaN   NaN   NaN  ...    RAYp  F2FyR07IdsN7I    am7c     NaN
NaN    NaN   NaN   NaN   NaN  ...    RAYp  F2FyR07IdsN7I     NaN     NaN
NaN    NaN   NaN   NaN   NaN  ...    RAYp  F2FyR07IdsN7I    mj86     NaN
NaN    NaN   NaN   NaN   NaN  ...    RAYp  F2FyR07IdsN7I    mj86     NaN

[5 rows x 229 columns]

In [3]:
# Step 3. Checking data types and missing values
print("Train data types:\n", train_data.dtypes)
print("Test data types:\n", test_data.dtypes)

print("Train data missing values:\n", train_data.isnull().sum())
print("Test data missing values:\n", test_data.isnull().sum())

Train data types:
 Var2      float64
Var3      float64
Var4      float64
Var5      float64
Var6      float64
           ...   
Var227     object
Var228     object
Var229     object
Var230    float64
y           int64
Length: 230, dtype: object
Test data types:
 Var2      float64
Var3      float64
Var4      float64
Var5      float64
Var6      float64
           ...   
Var226     object
Var227     object
Var228     object
Var229     object
Var230    float64
Length: 229, dtype: object
Train data missing values:
 Var2       9734
Var3       9734
Var4       9720
Var5       9759
Var6       1020
          ...  
Var227        0
Var228        0
Var229     5561
Var230    10000
y             0
Length: 230, dtype: int64
Test data missing values:
 Var2      2422
Var3      2422
Var4      2428
Var5      2439
Var6       271
          ... 
Var226       0
Var227       0
Var228       0
Var229    1423
Var230    2500
Length: 229, dtype: int64


In [4]:
# Step 4. Remove unnecessary columns and divide the data into features (X) and labels (y)
features_to_drop = ['Var8', 'Var15', 'Var20', 'Var31', 'Var32', 'Var39', 'Var42', 'Var48', 'Var52',
                    'Var55', 'Var79', 'Var141', 'Var167', 'Var169', 'Var175', 'Var185', 'Var209', 'Var230']
train_data.drop(columns=features_to_drop, inplace=True)
test_data.drop(columns=features_to_drop, inplace=True)

X = train_data.drop(columns=['y'])
y = train_data['y']

In [5]:
# Step 5. Using SimpleImputer, StandardScaler, and OneHotEncoder to process numeric and categorical data
numeric_features = X.select_dtypes(include=['int64', 'float64']).columns
categorical_features = X.select_dtypes(include=['object']).columns

numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

In [6]:
# Step 6. Combining numeric and categorical features
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)])

In [7]:
# Step 7. Building and training the model
model = RandomForestClassifier(random_state=42)

pipeline = ImbPipeline(steps=[
    ('preprocessor', preprocessor),
    ('smote', SMOTE(random_state=42)),
    ('classifier', model)])

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

pipeline.fit(X_train, y_train)

y_val_pred = pipeline.predict(X_val)

balanced_acc = balanced_accuracy_score(y_val, y_val_pred)
print(f'Balanced Accuracy: {balanced_acc:.2f}')

Balanced Accuracy: 0.68


In [8]:
# Step 8. Prediction and file creation

y_test_pred = pipeline.predict(test_data)

submission = pd.DataFrame({'y': y_test_pred})
submission.to_csv('submission.csv', index_label="index")

# Conclusions

***A value of 0.68 means that the model correctly classifies about 68% of the positive and negative classes in the validation dataset.***

## Strategies used to fill in missing values:

*There are several methods to recover data: mean, median, most_frequent, constant.*

- The median (‘median’) is the average value in an ordered data set that is not affected by extreme values, as can be the case with the mean, so it is the best choice for numeric features in this model.
- The use of a constant (‘constant’) to fill in missing values in categorical features allowed us to retain the information that the value was missing. This method allows you to fill in the missing values with a specific constant, such as ‘missing’, which makes it possible to clearly mark the missing values as a separate category. This can be useful for the model, as the missing data may have a specific meaning.

Also, when one of the categories (e.g., positive or negative class) in the dataset is significantly larger than the other, the model may refocus on the more frequent class, ignoring the less represented one. This leads to poor model performance on the less represented class, even if the overall accuracy looks high. Therefore, to eliminate this problem, SMOTE was used to create new data located on the segments between the points of the less represented class. This allowed us to ‘balance’ the distribution of classes and train the model to better recognise all categories.

## Results

*Model predictions are stored in the **submission.csv** file, which contains the predicted churn status for each client in the test set.*